File: backend/agents/cluster.py
"""
Cluster Agent
Groups posts into topic clusters using TF-IDF + KMeans.
Auto-detects optimal K using silhouette score.
Extracts cluster labels from top TF-IDF terms.
"""
import time
import re
from collections import Counter
from typing import List
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.cluster import KMeans, MiniBatchKMeans
from sklearn.preprocessing import normalize
from memory import SharedMemory, Cluster, ProcessedPost, AgentLog

STOPWORDS = {
    "the","a","an","and","or","but","in","on","at","to","for","of","with",
    "is","are","was","were","be","been","have","has","had","do","does","did",
    "will","would","could","should","this","that","these","those","i","we",
    "you","he","she","it","they","my","our","your","his","her","its","their",
    "get","got","just","like","really","very","also","even","still","back",
    "use","using","used","make","made","go","going","good","bad","new",
    "one","two","time","day","week","month","year","company","product",
    "people","way","know","want","need","can","said","dont","doesnt",
    "https","http","www","com","just","im","its"
}


def _clean(text: str) -> str:
    text = re.sub(r"https?://\S+", "", text.lower())
    text = re.sub(r"[^a-z0-9\s]", " ", text)
    tokens = [t for t in text.split() if len(t) > 2 and t not in STOPWORDS]
    return " ".join(tokens)


def _pick_k(X, min_k: int = 3, max_k: int = 12) -> int:
    """Choose optimal K using inertia elbow heuristic (fast, no silhouette cost)."""
    n = X.shape[0]
    max_k = min(max_k, n - 1)
    min_k = min(min_k, max_k)
    if min_k >= max_k:
        return min_k

    inertias = []
    ks = list(range(min_k, max_k + 1))
    for k in ks:
        km = MiniBatchKMeans(n_clusters=k, random_state=42, n_init=3)
        km.fit(X)
        inertias.append(km.inertia_)

    # Elbow: find biggest drop in inertia gradient
    if len(inertias) < 3:
        return min_k
    diffs = [inertias[i] - inertias[i + 1] for i in range(len(inertias) - 1)]
    best_idx = int(np.argmax(diffs))
    return ks[best_idx]


def _top_terms(posts: List[ProcessedPost], n: int = 5) -> List[str]:
    words = []
    for p in posts:
        words.extend(_clean(f"{p.title} {p.body}").split())
    freq = Counter(words)
    return [w for w, _ in freq.most_common(n)]


def run_cluster(memory: SharedMemory) -> None:
    start = time.time()
    posts = memory.processed_posts

    if not posts:
        memory.agent_logs.append(AgentLog("Cluster Agent", "error", "No posts to cluster", time.time()))
        return

    memory.agent_logs.append(AgentLog("Cluster Agent", "running", f"Vectorizing {len(posts)} posts with TF-IDF...", time.time()))

    texts = [_clean(f"{p.title} {p.body}") for p in posts]

    # TF-IDF vectorization
    vec = TfidfVectorizer(
        max_features=3000,
        ngram_range=(1, 2),
        min_df=2,
        sublinear_tf=True,
    )
    try:
        X = vec.fit_transform(texts)
    except ValueError:
        # If min_df=2 causes issues with small datasets
        vec = TfidfVectorizer(max_features=1000, ngram_range=(1, 2))
        X = vec.fit_transform(texts)

    X_norm = normalize(X)

    # Choose K
    n_docs = X_norm.shape[0]
    k = _pick_k(X_norm.toarray() if hasattr(X_norm, 'toarray') else X_norm, min_k=3, max_k=min(12, n_docs - 1))

    memory.agent_logs.append(AgentLog("Cluster Agent", "running", f"Running KMeans with K={k}...", time.time()))

    km = KMeans(n_clusters=k, random_state=42, n_init=10)
    labels = km.fit_predict(X_norm)

    # Get top terms per cluster from TF-IDF vocabulary
    feature_names = vec.get_feature_names_out()
    order_centroids = km.cluster_centers_.argsort()[:, ::-1]

    clusters = []
    for cluster_idx in range(k):
        # Posts in this cluster
        cluster_posts = [posts[i] for i, l in enumerate(labels) if l == cluster_idx]
        if not cluster_posts:
            continue

        # Top TF-IDF terms as cluster label
        top_terms = [feature_names[j] for j in order_centroids[cluster_idx][:6]]
        top_terms = [t for t in top_terms if t not in STOPWORDS][:5]
        label = " / ".join(top_terms[:3]) if top_terms else f"Topic {cluster_idx+1}"

        sentiments = [p.sentiment_score for p in cluster_posts]
        avg_sent = float(np.mean(sentiments))
        neg_count = sum(1 for p in cluster_posts if p.sentiment_label == "negative")
        negativity = neg_count / len(cluster_posts)

        subreddits = list({p.subreddit for p in cluster_posts if p.subreddit})
        sources = list({p.source for p in cluster_posts})
        times = [p.created_utc for p in cluster_posts]

        clusters.append(Cluster(
            id=f"cluster-{cluster_idx}",
            label=label,
            keywords=top_terms,
            posts=cluster_posts,
            volume=len(cluster_posts),
            avg_sentiment=round(avg_sent, 3),
            negativity=round(negativity, 3),
            subreddits=subreddits,
            sources=sources,
            first_seen=min(times),
            last_seen=max(times),
        ))

    # Sort by volume descending
    clusters.sort(key=lambda c: c.volume, reverse=True)
    memory.clusters = clusters

    duration = int((time.time() - start) * 1000)
    memory.agent_logs.append(AgentLog(
        "Cluster Agent", "done",
        f"Formed {len(clusters)} topic clusters (K={k}, TF-IDF vocab={X.shape[1]})",
        time.time(), duration
    ))

File: Insight/src/agents/cluster.ts
/**
 * Cluster Agent (Tier 1)
 * Groups posts by topical similarity using TF-IDF + cosine similarity.
 * Adapted from World Monitor's analysis.worker.ts Jaccard approach,
 * extended with TF-IDF term weighting for better topic separation.
 */
import type { ProcessedPost, Cluster } from '@/types';

function tokenize(text: string): string[] {
  const stopwords = new Set([
    'the','a','an','and','or','but','in','on','at','to','for','of','with','is',
    'are','was','were','be','been','being','have','has','had','do','does','did',
    'will','would','could','should','may','might','can','this','that','these',
    'those','i','we','you','he','she','it','they','my','our','your','his','her',
    'its','their','what','which','who','how','when','where','why','all','any',
    'both','each','few','more','most','other','some','such','no','nor','not',
    'only','own','same','so','than','too','very','just','as','if','about','after',
    'before','up','out','from','there','here','get','got','also','even','still',
    'back','then','now','like','well','into','use','make','made','want','need',
    'much','many','every','am',
  ]);
  return text.toLowerCase()
    .replace(/[^a-z0-9\s]/g, ' ')
    .split(/\s+/)
    .filter(t => t.length > 2 && !stopwords.has(t));
}

function buildTfIdf(docs: string[][]): Map<string, number[]> {
  const termDf = new Map<string, number>();
  for (const doc of docs) {
    const seen = new Set(doc);
    for (const term of seen) termDf.set(term, (termDf.get(term) ?? 0) + 1);
  }
  const N = docs.length;
  const tfidf = new Map<string, number[]>();

  docs.forEach((doc, i) => {
    const tf = new Map<string, number>();
    for (const term of doc) tf.set(term, (tf.get(term) ?? 0) + 1);

    for (const [term, count] of tf) {
      const tfScore = count / doc.length;
      const idf = Math.log(N / ((termDf.get(term) ?? 1) + 1)) + 1;
      const score = tfScore * idf;
      if (!tfidf.has(term)) tfidf.set(term, new Array(N).fill(0));
      tfidf.get(term)![i] = score;
    }
  });

  return tfidf;
}

function getDocVector(docIndex: number, tfidf: Map<string, number[]>): Map<string, number> {
  const vec = new Map<string, number>();
  for (const [term, scores] of tfidf) {
    if (scores[docIndex]! > 0) vec.set(term, scores[docIndex]!);
  }
  return vec;
}

function cosineSimilarity(a: Map<string, number>, b: Map<string, number>): number {
  let dot = 0, normA = 0, normB = 0;
  for (const [term, aVal] of a) {
    dot += aVal * (b.get(term) ?? 0);
    normA += aVal * aVal;
  }
  for (const bVal of b.values()) normB += bVal * bVal;
  const denom = Math.sqrt(normA) * Math.sqrt(normB);
  return denom === 0 ? 0 : dot / denom;
}

function getTopTerms(tokens: string[], n: number): string[] {
  const freq = new Map<string, number>();
  for (const t of tokens) freq.set(t, (freq.get(t) ?? 0) + 1);
  return [...freq.entries()].sort((a, b) => b[1] - a[1]).slice(0, n).map(e => e[0]);
}

export async function runClusterAgent(
  posts: ProcessedPost[],
  onProgress: (msg: string) => void,
): Promise<Cluster[]> {
  onProgress('Cluster Agent: Computing TF-IDF vectors...');
  await new Promise(r => setTimeout(r, 400));

  if (posts.length === 0) return [];

  const tokenized = posts.map(p => tokenize(p.cleanText));
  const tfidf = buildTfIdf(tokenized);
  const vectors = posts.map((_, i) => getDocVector(i, tfidf));

  const THRESHOLD = 0.18;
  const assigned = new Set<number>();
  const rawClusters: number[][] = [];

  for (let i = 0; i < posts.length; i++) {
    if (assigned.has(i)) continue;
    const cluster = [i];
    assigned.add(i);
    for (let j = i + 1; j < posts.length; j++) {
      if (assigned.has(j)) continue;
      const sim = cosineSimilarity(vectors[i]!, vectors[j]!);
      if (sim >= THRESHOLD) {
        cluster.push(j);
        assigned.add(j);
      }
    }
    rawClusters.push(cluster);
  }

  onProgress(`Cluster Agent: Formed ${rawClusters.length} raw clusters, building signals...`);
  await new Promise(r => setTimeout(r, 200));

  const clusters: Cluster[] = rawClusters
    .filter(c => c.length >= 1)
    .sort((a, b) => b.length - a.length)
    .slice(0, 12) // Top 12 clusters
    .map((indices, idx) => {
      const clusterPosts = indices.map(i => posts[i]!);
      const allTokens = clusterPosts.flatMap(p => tokenize(p.cleanText));
      const topTerms = getTopTerms(allTokens, 5);
      const label = topTerms.slice(0, 3).join(' / ') || `Topic ${idx + 1}`;
      const centroidText = clusterPosts[0]?.title ?? label;

      const avgSentiment = clusterPosts.reduce((s, p) => s + p.sentimentScore, 0) / clusterPosts.length;
      const negativePosts = clusterPosts.filter(p => p.sentimentLabel === 'negative');
      const negativity = Math.round((negativePosts.length / clusterPosts.length) * 100);
      const subreddits = [...new Set(clusterPosts.map(p => p.subreddit))];
      const timestamps = clusterPosts.map(p => p.created_utc);

      return {
        id: `cluster-${idx}`,
        label,
        posts: clusterPosts,
        centroidText,
        volume: clusterPosts.length,
        avgSentiment: Math.round(avgSentiment),
        negativity,
        subreddits,
        firstSeen: Math.min(...timestamps),
        lastSeen: Math.max(...timestamps),
      };
    });

  onProgress(`Cluster Agent: ${clusters.length} topic clusters identified ✓`);
  return clusters;
}
